### PR TITLE
Add native builds for ARM-based Macs

### PR DESCRIPTION
**Pull request by rtsao**
Resolves https://github.com/evanw/esbuild/issues/550

### CODE REVIEW

File: npm/esbuild-wasm/bin/esbuild [hunk 1]
```diff
@@ -81,4 +81,4 @@ fs.read = function () {
 };
 
 const argv = ['node', wasm_exec, esbuild_wasm].concat(process.argv.slice(2));
-wrapper(require, require.main, Object.assign(Object.create(process), { argv }), { instantiate });
+wrapper(require, require.main, Object.assign(Object.create(process), { argv }), Object.assign(Object.create(WebAssembly), { instantiate }));
```
Comment (rtsao) [inline npm/esbuild-wasm/bin/esbuild:84]: As of Golang 1.16.0, the generated `run` function now has this check:
```js
			if (!(instance instanceof WebAssembly.Instance)) {
				throw new Error("Go.run: WebAssembly.Instance expected");
			}
```

This seemed like the simplest solution as opposed to patching the generated code like in https://github.com/evanw/esbuild/blob/add2790afd5f72b39c7eea9e48b2022e56f4cfe3/scripts/esbuild.js#L61

File: lib/install.ts [hunk 2]
```diff
@@ -90,14 +90,6 @@ function validateBinaryVersion(binaryPath: string): void {
...
   } catch (err) {
-    if (platformKey === 'darwin arm64 LE')
-      throw new Error(`${err && err.message || err}
-
-This install script is trying to install the x64 esbuild executable because the
-arm64 esbuild executable is not available yet. Running this executable requires
-the Rosetta 2 binary translator. Please make sure you have Rosetta 2 installed
-before installing esbuild.
-`);
     throw err;
```
Comment (lukewarlow) [inline lib/install.ts:93]: This try catch is now redundant.

**Comment by evanw** [comment]
Thanks so much for putting this together.

However, it's missing the most important part: has anyone with a M1 processor tried this build command on Go 1.16.0 and checked that esbuild works? I don't want to ship a broken build if no one has tested it, but I don't have a M1 myself. A good check could be to build one of the benchmarks (e.g. `make bench-three-esbuild`).

I'm also curious to hear what the build times are like for native vs. Rosetta emulated. For example, what the build times are like for `make bench-three-esbuild` with `GOARCH=arm64` vs. `GOARCH=amd64`.

**Comment by sod** [comment]
> I'm also curious to hear what the build times are

m1 macbook pro 8gb | <- left arm64 ... right amd64 ->
image

**Comment by evanw** [comment]
> if I run the bench 10 times, the arm64 avg. at ~290ms and the amd64 at ~405ms

That's awesome. Thanks for checking.

**Comment by muuvmuuv** [comment]
Adding my tests here with Mac mini M1 16 GB Ram but with Go native arm64 build:

|amd64|arm64|
|-|-|
|Bildschirmfoto 2021-02-18 um 18 48 57|Bildschirmfoto 2021-02-18 um 18 49 46|

**Pull request merged**